Added List and Struct Encoding to arrow-avro Writer

jecsand838 opened the pull request:
# Which issue does this PR close?

- Part of https://github.com/apache/arrow-rs/issues/4886

# Rationale for this change

This refactor streamlines the `arrow-avro` writer by introducing a single, schema‑driven `RecordEncoder` that plans writes up front and encodes rows using consistent, explicit rules for nullability and type dispatch. It reduces duplication in nested/struct/list handling, makes the order of Avro union branches (null‑first vs null‑second) an explicit choice, and aligns header schema generation with value encoding. 

This should improve correctness (especially for nested optionals), make behavior easier to reason about, and pave the way for future optimizations. 

# What changes are included in this PR?

**High‑level:**

* Introduces a unified, schema‑driven `RecordEncoder` with a builder that walks the Avro record in Avro order and maps each field to its Arrow column, producing a reusable write plan. The encoder covers scalars and nested types (struct, (large) lists, maps, strings/binaries).
* Applies a single model of **nullability** throughout encoding, including nested sites (list items, fixed‑size list items, map values), and uses explicit union‑branch indices according to the chosen order.

**API and implementation details:**

* **Writer / encoder refactor**

  * Replaces the previous per‑column/child encoding paths with a **`FieldPlan`** tree (variants for `Scalar`, `Struct { … }`, and `List { … }`) and per‑site `nullability` carried from the Avro schema.
  * Adds encoder variants for `LargeBinary`, `Utf8`, `Utf8Large`, `List`, `LargeList`, and `Struct`. 
  * Encodes union branch indices with `write_optional_index` (writes `0x00/0x02` according to Null‑First/Null‑Second), replacing the old branch write.

* **Schema generation & metadata**

  * Moves the **`Nullability`** enum to `schema.rs` and threads it through schema generation and writer logic.
  * Adds `AvroSchema::from_arrow_with_options(schema, Option<Nullability>)` to either reuse embedded Avro JSON or build new Avro JSON that **honors the requested null‑union order at all nullable sites**. 
  * Adds `extend_with_passthrough_metadata` so Arrow schema metadata is copied into Avro JSON while skipping Avro‑reserved and internal Arrow keys.
  * Introduces helpers like `wrap_nullable` and `arrow_field_to_avro_with_order` to apply ordering consistently for arrays, fixed‑size lists, maps, structs, and unions. 
 
* **Format and glue**

  * Simplifies `writer/format.rs` by removing the `EncoderOptions` plumbing from the OCF format; `write_long` remains exported for header writing.

# Are these changes tested?

Yes.

* Adds focused unit tests in `writer/encoder.rs` that verify scalar and string/binary encodings (e.g., Binary/LargeBinary, Utf8/LargeUtf8) and validate length/branch encoding primitives used by the writer. 
* Round trip integration tests that validate List and Struct decoding in `writer/mod.rs`.
* Adjusts existing schema tests (e.g., decimal metadata expectations) to align with the new schema/metadata handling. 

# Are there any user-facing changes?

N/A because arrow-avro is not public yet.

jecsand838 commented on arrow-avro/src/codec.rs:
```diff
@@ -28,19 +28,6 @@ use std::borrow::Cow;
 use std::collections::HashMap;
 use std::sync::Arc;
 
-/// Avro types are not nullable, with nullability instead encoded as a union
-/// where one of the variants is the null type.
-///
-/// To accommodate this we special case two-variant unions where one of the
-/// variants is the null type, and use this to derive arrow's notion of nullability
-#[derive(Debug, Copy, Clone, PartialEq)]
-pub enum Nullability {
-    /// The nulls are encoded as the first union variant
-    NullFirst,
-    /// The nulls are encoded as the second union variant
-    NullSecond,
-}
```
I moved this to `src/schema.rs` to keep the imports clean and non-circular.

jecsand838 commented:
@scovich @alamb Would either of you be able to review this one? This should be the last large PR for `arrow-avro` as it lays the full foundation for the Writer. I tried to slim it down as much as possible.

nathaniel-d-ef commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -275,3 +384,434 @@ impl F64Encoder<'_> {
+    #[inline]
+    fn precomputed_union_value_branch(&self, order: Option<Nullability>) -> Option<u8> {
+        match (order, self.has_nulls()) {
+            (Some(Nullability::NullFirst), false) => Some(0x02), // value branch index 1
```
This is explained elsewhere, I think we can eliminate the comments. Maybe alias the values as constants globally.

nathaniel-d-ef commented on arrow-avro/src/writer/format.rs:
```diff
@@ -85,7 +66,7 @@ impl AvroFormat for AvroOcfFormat {
             None => "null",
         };
-        write_long(writer, 2)?;
+        write_long(writer, 2)?; // two entries
```
This note seems unnecessary

scovich commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -275,3 +384,434 @@ impl F64Encoder<'_> {
+
+    #[inline]
+    fn encode<W: Write + ?Sized>(&mut self, idx: usize, out: &mut W) -> Result<(), ArrowError> {
+        if let Some(b) = self.pre {
+            return out
+                .write_all(&[b])
+                .map_err(|e| ArrowError::IoError(format!("write union value branch: {e}"), e))
+                .and_then(|_| self.encode_inner(idx, out));
+        }
+        if let Some(order) = self.nullability {
+            let is_null = self.is_null(idx);
+            write_optional_index(out, is_null, order)?;
+            if is_null {
+                return Ok(());
+            }
+        }
+        self.encode_inner(idx, out)
```
Staring more at the code, I don't think `with_effective_nullability` should exist as a separate method. The `nullability` and `pre` are too closely tied up with `nulls` and `has_nulls`.

Also, there are too many Option here, which explodes the number of possible states we have to consider. Is there a way to simplify it?

|nulls|has_nulls|nullability|pre|note
|:-:|:-:|:-:|:-:|-
|None|true|*|*|**INVALID** -- `has_null` implies Some `nulls`
|*|true|None|*|**INVALID** -- `has_nulls` but writer is non-nullable
|*|*|None|Some|**INVALID** -- Some `pre` implies Some `nullability`
|*|true|*|Some|**INVALID** -- Some `pre` implies not `has_nulls`
|*|false|Some|None|**Missed optimization** -- we should have Some `pre`
|None|false|None|None|Everyone agrees this is a non-nullable column
|None|false|Some|Some|Non-nullable data with nullable writer (leverage `pre`)
|Some|false|None|None|Nullable data with no nulls and non-nullable writer
|Some|false|Some|Some|Nullable data with no nulls and nullable writer (leverage `pre`)
|Some|true|Some|None|Nullable data with nulls and nullable writer (cannot leverage `pre`)

scovich replied:
One thought that struck me this morning -- do we actually need `has_nulls` at all?
Or is it actually just an intermediate calculation? 

In other words, if the constructor sees `!has_nulls` then:
* set `nulls = None` even if the input array had a null buffer (because we just proved it's a no-op)
* generate `pre`

|nulls|nullability|pre|note
|:-:|:-:|:-:|-
|None|None|None|Non-nullable data with non-nullable writer
|None|None|Some|**INVALID** -- Some `pre` implies Some `nullability`
|None|Some|None|**Missed optimization** -- we should have Some `pre`
|None|Some|Some|Non-nullable data with nullable writer (leverage `pre`)
|Some|None|None|**INVALID** -- Some `nulls` but writer is non-nullable
|Some|None|Some|**INVALID** -- Some `pre` implies Some `nullability`<br>**INVALID** -- Some `nulls` but writer is non-nullable<br>**INVALID** -- Some `pre` implies None `nulls`
|Some|Some|None|Nullable data with nulls and nullable writer (cannot leverage `pre`)
|Some|Some|Some|**INVALID** -- Some `pre` implies None `nulls`

... but it's still a huge truth table with only three valid states...

Maybe better to just define an enum with three variants?
```rust
enum NullState {
    // non-nullable data with non-nullable writer
    NonNullable, 
    // non-nullable data with nullable writer
    NullableButNoNulls(NullOrdering, u8),
    // nullable data with nullable writer
    Nullable(NullBuffer, NullOrdering),
}
```

scovich reviewed (commented):
Flushing a partial review. I got all the files except encoder.rs, which is partially reviewed.

jecsand838 commented on arrow-avro/src/schema.rs:
```diff
@@ -370,6 +371,49 @@ impl AvroSchema {
+        let record_name = schema
+            .metadata
+            .get(AVRO_NAME_METADATA_KEY)
+            .map_or("topLevelRecord", |s| s.as_str());
```
> aside: Is this a well-known default name? Or just an arbitrary naming choice by this package?

While not an Avro‑spec default, `topLevelRecord` is used as a de‑facto default because several popular tools (notably Spark/Databricks) default to the same name when they synthesize an Avro schema from a struct/row.

> And does it actually matter in practice? (I guess if it mattered, the schema metadata would say so)?

Avro requires that a record have a name to be valid. Also because the record name participates in canonical form parsing, changing a record's name will change it's fingerprint.

jecsand838 commented on arrow-avro/src/writer/mod.rs:
```diff
@@ -63,19 +66,42 @@ impl WriterBuilder {
     {
-        Writer {
+        let mut format = F::default();
+        let avro_schema = if let Some(json) = self.schema.metadata.get(SCHEMA_METADATA_KEY) {
+            AvroSchema::new(json.clone())
+        } else {
+            AvroSchema::try_from(&self.schema)?
+        };
```
That is much cleaner and idiomatic.

jecsand838 commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -84,48 +85,211 @@ fn write_bool<W: Write + ?Sized>(writer: &mut W, v: bool) -> Result<(), ArrowErr
+/// Per‑site encoder plan for a field. This mirrors Avro structure so nested
+/// optional branch order can be honored exactly as declared by the schema.
+#[derive(Debug, Clone)]
+enum FieldPlan {
```
@scovich I have additional plans for maps, enums, decimals, etc on another branch. I just removed them from this PR to reduce the size.

jecsand838 commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -84,48 +85,211 @@ fn write_bool<W: Write + ?Sized>(writer: &mut W, v: bool) -> Result<(), ArrowErr
-    write_int(writer, branch)
+    // For NullFirst: null => 0x00, value => 0x02
+    // For NullSecond: value => 0x00, null => 0x02
+    let byte = match order {
+        Nullability::NullFirst => {
+            if is_null {
+                0x00
+            } else {
+                0x02
+            }
+        }
+        Nullability::NullSecond => {
+            if is_null {
+                0x02
+            } else {
+                0x00
+            }
+        }
```
I like that third approach. Also @nathaniel-d-ef recommended adding a constant for `0x00` and `0x02` to improve readability.

jecsand838 commented on arrow-avro/src/schema.rs:
```diff
@@ -790,12 +856,35 @@ fn merge_extras(schema: Value, mut extras: JsonMap<String, Value>) -> Value {
+    }
+}
+
 fn datatype_to_avro(
```
> can we keep the nice name and just add the extra arg

This is the way. I'll makes those changes. You're 100% correct.

scovich commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -275,3 +384,434 @@ impl F64Encoder<'_> {
+        })
+    }
+
+    #[inline]
```
I wouldn't be shocked if one or two inline markers actually matter... but no idea which two. Can we just remove all the markers and track a follow-up issue to make profile-guided decisions on which ones to bring back?

scovich replied:
I bet profile-guided optimizations would give back far more than 4%, since it covers far more than just inlining decisions. If we care about optimizing that much, we should probably go that route? But that requires benchmarking and profiling infra that arrow-rs probably lacks. An hybrid (and probably almost optimal) approach would be to manually gather .profdata, and then see what `llvm-profdata show` has to say about these functions. 

Humans are provably terrible at guessing where to sprinkle hints like inlining or branch prediction, and so I would expect that some of the inline markers are badly-placed and actually _hurting_ performance, which eats into the performance gains of the well-placed markers. Automatic or manual profile-guided optimization at least gives some concrete data to replace those guesses.

Overall, my recommended approach would be to remove all the inline markers for now (focus on getting correct code in), and if we want to chase the 4% that could be a follow-up item to do it right. Preferably after the code has stabilized, since it's still evolving rapidly right now.

jecsand838 replied:
100% fair point. I went ahead and removed them.

jecsand838 commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -84,36 +81,186 @@ fn write_bool<W: Write + ?Sized>(writer: &mut W, v: bool) -> Result<(), ArrowErr
+enum NullState {
+    NonNullable,
+    NullableNoNulls {
+        byte: u8,
```
100%, much more descriptive.

jecsand838 commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -83,49 +81,358 @@ fn write_bool<W: Write + ?Sized>(writer: &mut W, v: bool) -> Result<(), ArrowErr
+            }
+        };
+        let mut columns = Vec::with_capacity(root_fields.len());
+        for root_field in root_fields.iter() {
```
You're right, this is a good catch.

jecsand838 commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -267,11 +513,275 @@ impl F32Encoder<'_> {
+    // Emit a single positive block for performance, then the end marker.
+    write_long(out, len as i64)?;
+    for j in start..end {
+        write_item(j, out)?;
```
That's very fair, I'll clean that up.

jecsand838 commented:
> Bunch of nits to simplify the code, but otherwise ready to merge.
> 
> > There are two items I wanted to follow-up on in future PRs:
> > ```
> > 1. Further `arrow-avro/src/schema.rs` file refactoring (if possible). As I updated the file I saw other potential areas to improve. For example, I think if we introduced some sort of struct to hold context for `datatype_to_avro`, we could simplify and improve the logic.
> > ```
> 
> Definitely future work.
> 
> > ```
> > 2. Renaming `Nullability`. The name change will impact other areas of the code not related to the `Writer`. I'd also want to consider the impact `Union` type support will have as well when we get there. But I do think @scovich has a valid point about the current naming convention.
> > ```
> 
> Yes. Mixing refactors with new functionality is always messy.
> 
> I would recommend two other follow-ups: 3. Profile-guided placement of `#[inline]` markers 4. Harmonize arg ordering for the various encoding methods that take `out`; some pass it as the first arg, others as the second arg.

@scovich All great call outs. I accepted your latest suggestions and pushed up changes that harmonized the arg ordering for the methods taking `out` as a parameter. As always thank you for the solid and thorough review 😄. We'll also be sure to follow-up on those PRs.

@alamb @mbrobbel Would either of you be able to look over this PR if you get a chance?

scovich commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -83,181 +78,417 @@ fn write_bool<W: Write + ?Sized>(writer: &mut W, v: bool) -> Result<(), ArrowErr
+        plan: &FieldPlan,
+        nullability: Option<Nullability>,
+    ) -> Result<Self, ArrowError> {
+        let has_nulls = array.null_count() > 0;
```
```suggestion
```
no longer used?

jecsand838 replied:
Ty for catching that. It's definitely no longer being used.

mbrobbel commented on arrow-avro/src/writer/encoder.rs:
```diff
@@ -83,181 +74,416 @@ fn write_bool<W: Write + ?Sized>(writer: &mut W, v: bool) -> Result<(), ArrowErr
+                }
+            },
+            other => {
+                return Err(ArrowError::NotYetImplemented(
+                    "Avro writer: {other:?} not yet supported".into(),
+                ));
```
```suggestion
                return Err(ArrowError::NotYetImplemented(format!(
                    "Avro writer: {other:?} not yet supported".into(),
                )));
```

alamb commented:
Thanks @jecsand838
